src/CountIterableCapableTrait.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Dhii\Iterator;
+
+use Traversable;
+use IteratorAggregate;
+use Countable;
+
+/**
+ * Functionality for counting elements in an iterable.
+ *
+ * @since [*next-version*]
+ */
+trait CountIterableCapableTrait
+{
+    /**
+     * Counts the elements in an iterable.
+     *
+     * Is optimized to retrieve count from values that support it.
+     * - If array, will count in regular way using count();
+     * - If {@see Countable}, will do the same;
+     * - If {@see IteratorAggregate}, will drill down into internal iterators
+     * until the first {@see Countable} is encountered, in which case the same
+     * as above will be done.
+     * - In any other case, will apply {@see iterator_count()}, which means
+     * that it will iterate over the whole traversable to determine the count.
+     *
+     * @since [*next-version*]
+     *
+     * @param array|Traversable $iterable The iterable to count. Must be finite.
+     *
+     * @return int The amount of elements.
+     */
+    protected function _countIterable($iterable)
+    {
+        $resolve = function ($it) {
```

Comment:
Why not separate this closure into a separate trait method? I usually like putting functionality like this inside closures; it's clean and very readable. However, I think this particular case merits a separate method. Something like `_resolveIterator()`.

This would also allows the resolution functionality to be overridden and to be tested separately.

Reply:
@mecha, I started by having that method, with that exact name, `_resolveIterator()`. However, the more I thought, the more I came to the conclusion that this resolution is very specific, because it specifically resolves to a `Countable|null`. A generic resolution mechanism would have to accept a `callable` either way, which would test the resolved `Traversable`, and stop if match found. Another option would be to have a `_resolveCountableIterator()`. Thoughts?

Reply:
Sure, `_resolveCountableIterator()` sounds good to me. A generic resolution mechanism would not be suitable here, since as you said, you specifically need a `Countable` or a `null` result.

But having something which given an iterator gives you a countable is just what you need.

It would also improve the test for this trait, allowing you to mock the return of `_resolveCountableIterator()` and test `_countIterable()` in isolation, while also testing `_resolveCountableIterator()`. This, I think, is better than mocking nested hierarchies of iterators to test the counting; such mocking should be done to test the resolution, imo.

Reply:
I still don't like the idea. I think this logic is very specific to this particular method. Abstracting it would require the implementing class to implement the non-generic resolution. But this resolution isn't even required for all cases. I would really prefer to leave it this way, at least for now.

Reply:
I agree with your concerns about burden on implementing classes. However, consider simply making the closure its own non-abstract method. There is nothing to lose and a lot of testing benefits to be had, as well as cleaner code and the ability to override.